services/construction.go

```diff
@@ -124,3 +145,713 @@ func (s *constructionAPIService) ConstructionSubmit(
+	return readCurrency(amount, OasisCurrency, true)
+}
+
+func readPoolShareNeg(amount *types.Amount) (*quantity.Quantity, error) {
+	return readCurrency(amount, PoolShare, true)
+}
+
+// ConstructionPayloads implements the /construction/payloads endpoint.
+func (s *constructionAPIService) ConstructionPayloads(
+	ctx context.Context,
+	request *types.ConstructionPayloadsRequest,
+) (*types.ConstructionPayloadsResponse, *types.Error) {
+	terr := ValidateNetworkIdentifier(ctx, s.oasisClient, request.NetworkIdentifier)
+	if terr != nil {
+		loggerCons.Error("ConstructionPayloads: network validation failed", "err", terr.Message)
+		return nil, terr
+	}
+
+	// Payloads is called with an array of operations and the response from
+	// `/construction/metadata`. It returns an unsigned transaction blob and
+	// a collection of payloads that must be signed by particular addresses
+	// using a certain SignatureType. The array of operations provided in
+	// transaction construction often times can not specify all "effects" of
+	// a transaction (consider invoked transactions in Ethereum). However,
+	// they can deterministically specify the "intent" of the transaction,
+	// which is sufficient for construction. For this reason, parsing the
+	// corresponding transaction in the Data API (when it lands on chain)
+	// will contain a superset of whatever operations were provided during
+	// construction.
+
+	nonceRaw, ok := request.Metadata[NonceKey]
+	if !ok {
+		loggerCons.Error("ConstructionPayloads: nonce metadata not given")
+		return nil, ErrMalformedValue
+	}
+	nonceF64, ok := nonceRaw.(float64)
+	if !ok {
+		loggerCons.Error("ConstructionPayloads: malformed nonce metadata")
+		return nil, ErrMalformedValue
+	}
+	nonce := uint64(nonceF64)
+
+	if len(request.Operations) < 2 {
+		loggerCons.Error("ConstructionPayloads: missing fee operation")
+		return nil, ErrMalformedValue
+	}
+	feeOp := request.Operations[0]
+	if feeOp.Type != OpTransfer {
+		loggerCons.Error("ConstructionPayloads: fee operation wrong type",
+			"type", feeOp.Type,
+			"expected_type", OpTransfer,
+		)
+		return nil, ErrMalformedValue
+	}
+	if feeOp.Account.SubAccount == nil {
+		loggerCons.Error("ConstructionPayloads: missing fee operation subaccount")
+		return nil, ErrMalformedValue
+	}
+	if feeOp.Account.SubAccount.Address != SubAccountGeneral {
+		loggerCons.Error("ConstructionPayloads: fee operation wrong subaccount address",
+			"subaccount", feeOp.Account.SubAccount.Address,
+			"expected_subaccount", SubAccountGeneral,
+		)
+		return nil, ErrMalformedValue
+	}
+	signWithAddr := feeOp.Account.Address
+	feeAmount, err := readOasisCurrencyNeg(feeOp.Amount)
+	if err != nil {
+		loggerCons.Error("ConstructionPayloads: readOasisCurrency feeOp.Amount",
+			"amount", feeOp.Amount,
+			"err", err,
+		)
+		return nil, ErrMalformedValue
+	}
+
+	var method transaction.MethodName
+	var body cbor.RawMessage
+	if len(request.Operations) == 3 &&
```

Comment:
Could we change this to a switch statement, it seems more idiomatic.

Reply:
no? it needs to check multiple things

Reply:
Why wouldn't something like this work:
```golang
switch {
case len(request.Operations) == 3 && xxx:
// ...
case len(request.Operations) == 2 && xxx:
// ...
default:
// ...
}
```

Reply:
oh, I hadn't even notice that they have that syntax. fine I guess